### PR TITLE
feat: nft dutch auction

### LINKED ISSUE #187: Feature: NFT Dutch Auction
### Description

In this example, we will show the implementation of an NFT minting contract using a Dutch Auction, where the minting price follows a decreasing price curve.

### Minting Details

- The seller of the NFT deploys this contract with a `startingPrice`.
- The auction lasts for a specified `duration`.
- The price decreases over time.
- Participants can purchase NFTs at any time as long as the `totalSupply` has not been reached.
- The auction ends when either the `totalSupply` is reached or the `duration` has elapsed.

### Criteria:

- Price is denominated in `IERC20 priceToken`
- Include comprehensive tests
- Deploy the contract on a testnet

### Resources:

- [Solidity By Example - Dutch Auction](https://solidity-by-example.org/app/dutch-auction/)

### ODHack

To be eligible for additional rewards, be sure to review and follow the [ODHack Common Guidelines](https://www.notion.so/ODHack-Common-Guidelines-b9c6b6a4ac4146d087185568aca38a3f?pvs=21) and [Contributing Guidelines](https://github.com/NethermindEth/StarknetByExample/blob/main/CONTRIBUTING.md#contributing).
Be sure to join the telegram group and introduce yourself.

### PR DESCRIPTION
**Pull request by princeibs**
**Issue(s):** Close #187 

### Description

In this example, we show the implementation of an NFT minting contract using a Dutch Auction, where the minting price follows a decreasing price curve.

### Checklist

- [x] **CI Verifier:** Run `./scripts/cairo_programs_verifier.sh` successfully
- [x] **Contract Tests:** Added tests to cover the changes

### CODE REVIEW

File: Scarb.toml [hunk 1]
```diff
@@ -3,20 +3,20 @@ members = [
...
 test = "$(git rev-parse --show-toplevel)/scripts/test_resolver.sh"
 
-# [workspace.tool.snforge]
+[workspace.tool.snforge]
 
 [workspace.dependencies]
 starknet = ">=2.6.3"
-# snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry.git", tag = "v0.11.0" }
-openzeppelin = { git = "https://github.com/OpenZeppelin/cairo-contracts.git", tag="v0.11.0" }
+snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry.git", tag = "v0.23.0" }
```
Comment (julio4) [inline Scarb.toml:16]: Move forge to dev dependencies, also we will use version `0.24.0`

File: listings/applications/nft_auction/src/erc20.cairo
Comment (julio4) [inline listings/applications/nft_auction/src/erc20.cairo:1]: We have an implementation of an [ERC20](https://github.com/NethermindEth/StarknetByExample/tree/main/listings/applications/erc20) already, you can remove this and use it instead.

File: listings/applications/nft_auction/src/nft_auction.cairo
Comment (julio4) [inline listings/applications/nft_auction/src/nft_auction.cairo:1]: Can you define error codes in a specific module

File: listings/applications/nft_auction/src/nft_auction.cairo [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    #[abi(embed_v0)]
+    impl NFTAuction of super::INFTAuction<ContractState> {
+        fn get_price(self: @ContractState) -> u64 {
+            let time_elapsed = (get_block_timestamp() - self.start_at.read())
+                / 1000; // Ignore milliseconds
+            let discount = self.discount_rate.read() * time_elapsed;
+            let price: u64 = self.starting_price.read() - discount;
+            price
```
Comment (julio4) [inline listings/applications/nft_auction/src/nft_auction.cairo:102]: just return `self.starting_price.read() - discount`

Comment (julio4) [reply listings/applications/nft_auction/src/nft_auction.cairo:102]: Can you rename all the `nft_auction` by `nft_dutch_auction` (files, modules, etc...)

**Review by julio4 — CHANGES_REQUESTED** [review]
LGTM! I requested some slight changes, thank you!

File: Scarb.toml (hunk 1 above)
Comment (princeibs) [reply Scarb.toml]: moving snforge to dev dependencies throws this error 
image

**Review by julio4 — APPROVED** [review]
LGTM!

**Pull request merged**